src/server/generic_family.cc

```diff
@@ -463,6 +458,209 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
   (*cntx)->SendLong(match_cnt);
 }
 
+struct SortEntryEmpty {};
+
+struct SortEntryScore {
+  double score;
+};
+
+template <
+    // Whether alphanumeric sort will be used.
+    bool ALPHA,
+    // Deduced StorageKind type for optimal size.
+    container_utils::StorageKind STORAGE_KIND,
+    // Whether to store double score. It will be stored for non-alphanumeric sorts
+    // when our items are not trivailly convertiable to doubles (i.e. they are not integers)
+    bool STORE_SCORE = !ALPHA && STORAGE_KIND != container_utils::StorageKind::INTEGRAL>
+struct SortEntry : public std::conditional_t<STORE_SCORE, SortEntryScore, SortEntryEmpty> {
+  // Key type. If we sort alphanumerically, we store strings. Else we mimic
+  // the container storage type.
+  using KEY_TYPE =
+      std::conditional_t<ALPHA, std::string, container_utils::StorageType<STORAGE_KIND>>;
+  // Comparison type. std:string for alphanumeric sort, double otherwise.
+  using COMPARE_TYPE =
+      std::conditional_t<ALPHA, const std::string&,
+                         // If we store only integers, we can use them for comparison.
+                         std::conditional_t<std::is_integral_v<KEY_TYPE>, KEY_TYPE, double>>;
+
+  // This is either: std::string, int64_t or std::variant<std::string, int64_t>
+  KEY_TYPE key;
+
+  bool Parse(std::string&& item) {
+    if constexpr (STORE_SCORE) {
+      if (!absl::SimpleAtod(item, &this->score))
+        return false;
+    }
+    if constexpr (std::is_assignable_v<KEY_TYPE&, std::string&&>) {
+      key = std::move(item);
+    } else {
+      VLOG(0) << "Got string when expected int";
+      assert(false && "unreachable");
+    }
+    return true;
+  }
+
+  bool Parse(int64_t item) {
+    if constexpr (STORE_SCORE) {
+      this->score = item;
+    }
+    if constexpr (std::is_assignable_v<KEY_TYPE&, int64_t>) {
+      key = item;
+    } else {
+      VLOG(0) << "Got string when expected string";
+      assert(false && "unreachable");
+    }
+    return true;
+  }
+
+  COMPARE_TYPE Cmp() const {
+    if constexpr (STORE_SCORE) {
+      return this->score;
+    } else {
+      return key;
+    }
+  }
+
+  template <typename F> void visit(F&& f) const {
+    // TODO: better check
+    if constexpr (!ALPHA && STORAGE_KIND == container_utils::StorageKind::MIXED)
+      std::visit(std::forward<F>(f), key);
+    else
+      f(key);
+  }
+};
+
+namespace helper {
+// Sort entry indices are encoded with the formula ALPHA * 3 + STORAGE
+template <size_t I, bool ALPHA = I / 3, size_t STORAGE = I % 3> static auto MapType() {
+  return SortEntry<ALPHA, container_utils::StorageKind{STORAGE}>{};
+}
+
+template <size_t... seq> static auto ListTypes(std::index_sequence<seq...>) {
+  return std::variant<std::vector<decltype(MapType<seq>())>...>{};
+}
+}  // namespace helper
+
+// std::variant of all possible vector<SortEntry<*, *>>
+using SortEntryList = decltype(helper::ListTypes(std::make_index_sequence<6>{}));
```

Comment:
Tricky but works... 😕 

The real issue is that if someone decides to add some more enum variants, then this will break easily and silently. Maybe add a few static_asserts on the constants. AFAIK there is no way to get the enum size (without adding a `__LAST` member) 🤔

Reply:
I would like to understand in more detail what you do here. 
what's this vector does.

Reply:
I collected all possible vector<SortEntry<ALPHA, STORAGE>> instantiations into a `std::variant<>`

Reply:
you try to write a generic code that manages "generic" data structures and by generic I mean - union of vectors.
You say your type space is ALPHAxSTORAGE, but 
1. can ALPHA=TRUE and STORAGE be non-string? 
2. when ALPHA is FALSE you have list data structure or set with non-int encoding, your storage is STRING? 

what I am trying to say is that your contention point is probably the sorting algorithm where you want to have the most optimized code working on a compile-known type. Everything else can be deduced at run-time. So why not decompose your states in such way that they will lead into 3 branches: sort by string, sort by int and sort by MIXED. But once you are inside the branch you just write the code for that branch. 
Maybe I am oversimplifying things - tell me if I am wrong..

Reply:
In any case, the sequence generation is not justified here.
It's artificial in order to save a few lines. Lets use the explicit variant with comments mapping each type to a combination of SORT sub-case.  I am still struggling to see whether you really need the whole cartesian product of possibilities.